Adds info about creating a maliput_backend using maliput_sparse. (#126)
maliput/maliput_documentation: docs/tutorials/maliput_sparse/creating_maliput_backend.rst

Creating a maliput backend using maliput_sparse
===============================================

Context
-------

`maliput_sparse`_ is a convenient package that provides several helpers for creating a maliput backend that is expected to be built on top of waypoints without any analytical model of the surface.

By using the `builder API`_, the mathematical model is solved under the hood so the user doesn't have to dive into complex geometric calculations.

Creating a maliput backend
--------------------------

> 1. Implement a `maliput_sparse::parser::Parser`_: This pure virtual class is responsible for parsing an user-defined underlying road description format and serve the parsed data.

Reviewer: ```suggestion
1. Implement a `maliput_sparse::parser::Parser`_: This abstract class is responsible for parsing an user-defined underlying road description format and serve the parsed data.
```

>    During the parsing process, the user will have to fill some data structures about the parsed lanes and connections between lanes, that will be used by the `maliput_sparse`_'s builder to create the maliput backend.

Reviewer: ```suggestion
   During the parsing process, the user will have to fill some data structures about the parsed lanes and connections between them, that will be used by the `maliput_sparse`_'s builder to create the maliput backend.
```

  .. code-block:: cpp
    :linenos:

     class MyParser : public Parser {
      public:
       MyParser(const std::string& my_file) : Parser() {
        // TODO:
        // Parse underlying file and fill up junctions and connections.
       }
      private:
       const std::unordered_map<Junction::Id, Junction>& DoGetJunctions() const override {
         return junctions_;
       }
       const std::vector<Connection>& DoGetConnections() const override {
         return connections_;
       }

>       };

Reviewer: Nit: I would suggest adding `std::vector<Connection> connections_;` and `std::unordered_map<Junction::Id, Junction> junctions_;` as class attributes to make it super clear that those are not magic attributes that come from `Parser`. Probably, it is worth adding in the constructor a reference to populate them there.

Author: You're right. Done

  As example: Take a look at the `maliput_osm`'s parser implementation: `maliput_osm::osm::OSMManager`_.

2. Once the above class is implemented, the user can make use of the `maliput_sparse::loader::RoadNetworkLoader`_ to load the road network from the user-defined format by injecting the previously implemented `maliput_sparse::parser::Parser`_ class.

  .. code-block:: cpp
    :linenos:

      maliput_sparse::loader::BuilderConfiguration maliput_sparse_config;
      // ..
      // Fill up maliput_sparse_config with the desired parameters.
      //
      const std::string my_file{"my_file.txt"};
      std::unique_ptr<maliput_sparse::parser::Parser> my_parser =
            std::make_unique<my_parser::MyParser>(my_file);
      std::unique_ptr<maliput::api::RoadNetwork> road_network = maliput_sparse::loader::RoadNetworkLoader(std::move(my_parser), maliput_sparse_config)();

  As example: Take a look at the `maliput_osm`'s road network loader: `maliput_osm::builder::RoadNetworkBuilder`_.

> 3. Finally, it is recommended to provide a maliput::plugin::RoadNetworkLoader implementation with this new backend, in a way that it can be loaded in runtime to be used by packages like `maliput_viz`.

Reviewer: ```suggestion
3. Finally, it is recommended to provide a `maliput::plugin::RoadNetworkLoader` implementation with this new backend, in a way that it can be loaded in runtime to be used by packages like `maliput_viz`.
```
